### ISSUE TITLE
Question on Tikz and Alt Text

### ISSUE THREAD

**Issue by cfroccajr** [brucemiller/LaTeXML#2809]
The example below will create a document in which the tikzpicture has alt text.  When compiled using LuaLatex (2025) on Overleaf it produces a pdf with alt text for the tikzpicture.  Now that this is possible when compiling from latex to pdf is it possible that alt text for tikz will be supported in the next release of latexml.

`\DocumentMetadata{
  lang        = en-US,
  testphase={phase-III,table,math,firstaid},
  tagging-setup={math/setup={mathml-SE,mathml-AF}},
  tagging=on,
  pdfversion=2.0,
  pdfstandard = ua-2,
  pdfstandard = a-4f, %or a-4
  uncompress,
}
\AddToDocumentProperties[hyperref]{pdftitle}{Minimal Tagged Tikz}
\AddToDocumentProperties[hyperref]{pdfauthor}{Chuck Rocca}

\documentclass{article}
\usepackage{tikz}

\begin{document}

\begin{tikzpicture}[
    alt={a red square and a blue circle}
]
    \draw[fill=red] (0,0) rectangle (2,2);
    \draw[fill=blue] (4,1) circle (1cm);
\end{tikzpicture}

\end{document}
`

**Comment by dginev** [comment]
This is an excellent request and thank you for making it.

Scheduling releases is one of the great mysteries of LaTeXML life, but this kind of enhancement will definitely be high on our radar.

**Comment by teepeemm** [comment]
As with #2679, TikZ also takes `actualtext` and `artifact`.  Also the `{picture}` environment takes `alt`, `actualtext`, and `artifact`.

**Comment by brucemiller** [comment]
Oooh, thanks for that comment @teepeemm ; *very* timely!

**Comment by c-minz** [comment]
This is a comment on another preparation of alt-text for future support.

In the TikZ-based package `causets`, I implemented an `alt` key. Currently, the provided value is only stored in an internal macro but not processed further,
```tex
alt/.estore in=\causetAltText
```
For example,
```tex
\pcauset[alt={N poset}]{2,4,1,3}
```
is producing a small (inline) TikZ diagram, which could be replaced by the text "N poset" if the graphic cannot be shown (or a screen reader is used).

For the next version of the `causets` package, I will implement a forwarding of the alt-value in the context of the described issue above. Though at the moment, I believe that TikZ itself does not yet support alt-text.

**Comment by xworld21** [comment]
Drive by comment: SVG uses the `<svg:title>` and `<svg:description>` tags to embed alternative text. It appears to be good practice to use `aria-labelledby`, `aria-describedby` to link the SVG with title and description, as some screen readers don't pick it up. This is another case where you'll need to make an opinionated decision about where to map actualtext, alt.

**Comment by teepeemm** [comment]
> Though at the moment, I believe that TikZ itself does not yet support alt-text.

The commands `\tikzpicture` and `\includegraphics` and the environments `{tikzpicture}` and `{picture}` all support `alt`, `actualtext`, and `artifact` in LaTeX.  Or were you talking about LaTeXML?

**Comment by c-minz** [comment]
I have to admit that I did not consult the manual of [graphicx](https://ctan.org/pkg/graphicx) for a long time, so I was not aware of the already existing accessibility support with the keywords `alt`, `actualtext` and `artifact`. (This is great news for my next [causets](https://ctan.org/pkg/causets) package update, because this has the option to externalize TikZ-pictures with `\usetikzlibrary{external}` and handle reuse of such pictures with `\includegraphics`).

However, I am not aware of standard TikZ supporting such keywords out of the box, at least the [TikZ documentation](https://tikz.dev/) does not mention this anywhere. There is the somewhat independent [**latex-lab** project](https://latex.org.uk/macros/latex/required/latex-lab/), which offers such tagging features when using code like `\DocumentMetadata{... tagging=on, ...}` as in the initial example above. I am considering to forward data to the keywords `alt`, `actualtext` and `artifact` in this extended context, but it would be even better if these indeed come already with standard TikZ. @teepeemm Are these already supported in standard TikZ and where is this documented?

**Comment by teepeemm** [comment]
@c-minz  I'm having trouble understanding your use case.  You are correct that TikZ does not support `alt` on its own, but does if `\DocumentMetadata` was used.  But if you are aiming for accessibility, then you need `\DocumentMetadata` (and LuaLaTeX) anyway. This loads the latex-lab project (see also https://latex3.github.io/tagging-project/), which is independent of TikZ I guess, but definitely not independent of the rest of LaTeX.

Or are you needing to determine if `\DocumentMetadata` was used, so that you know whether to forward the key or not?  In that case, you want to look into `\IfDocumentMetadataTF` (and `...T` and `...F`), see `texdoc ltnews35`.

**Comment by dginev** [comment]
Sorry to jump in, but I think there should be a gentle correction here:

> But if you are aiming for accessibility, then you need \DocumentMetadata (and LuaLaTeX) anyway.

My current understanding of the legal regulations is that **one** accessible format needs to be made available for all materials. If someone is investing effort to produce accessible HTML5 with LaTeXML, the accessible PDF route is a "nice to have" and not a "need". I would make the symmetrical statement if one already had accessible PDF 2.0 via lualatex.

**Comment by e-duckworth** [comment]
I would really love for a work around solution to this before 0.9.1 comes out (not that I have any standing or desire to pressure the volunteers making it happen).  The whole reason I started using latexml (maybe large reason for its development?) was to create documents that are more accessible. This seems like a fairly large gap in that goal. In the meantime I'm considering doing this, but it's much much worse than a real solution
```
\iflatexml
\includegraphics[alt = {a line}]{foobar}
\else
\tikzexternalize
\tikzsetnextfilename{foobar}
\begin{tikzpicture}[alt = {a line}]
\draw (0,0) -- (1,0);
\end{tikzpicture}
\fi
```

**Comment by c-minz** [comment]
@teepeemm Sorry for my late response. Like @dginev wrote, I believe the accessibility support should be independent for LaTeXML and PDF. Since some LaTeX package (like graphicx and causets) have keywords like alt, these could be supported by LaTeXML independently. LuaLaTeX is surely a great option to have, but not everyone is using it, especially for articles on the arXiv (where I found my initial contact with the LaTeXML project).

So perhaps there are other packages as well, but here are those that offer some options already to be picked up:
- graphicx with keyword `alt`, `actualtext` and `artifact` in the optional parameter of `\includegraphics` ...
- causets with keyword `alt` as described above (I'm also considering to add the keywords `actualtext` and `artifact` for the next package release)
